packit/sync.py

```diff
@@ -1,127 +1,144 @@
-#
-# Copyright (c) 2019 Red Hat, Inc.
-
-# Permission is hereby granted, free of charge, to any person obtaining a copy
-# of this software and associated documentation files (the "Software"), to deal
-# in the Software without restriction, including without limitation the rights
-# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
-# copies of the Software, and to permit persons to whom the Software is
-# furnished to do so, subject to the following conditions:
-#
-# The above copyright notice and this permission notice shall be included in all
-# copies or substantial portions of the Software.
-#
-# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-# SOFTWARE.
+# Copyright Contributors to the Packit project.
+# SPDX-License-Identifier: MIT
 
+import glob
 import logging
-import shutil
 from pathlib import Path
-from typing import List
-from typing import NamedTuple, Union
+from typing import List, Optional, Union, Sequence
 
 from packit.exceptions import PackitException
+from packit.utils import run_command
 
 logger = logging.getLogger(__name__)
 
 
-class SyncFilesItem(NamedTuple):
-    src: Union[str, List[str]]
-    dest: str
+def check_subpath(subpath: Path, path: Path) -> Path:
+    """Check if 'subpath' is a subpath of 'path'
+    Args:
+        subpath: Subpath to be checked.
+        path: Path agains which subpath is checked.
+    Returns:
+        'subpath', in case it is a subpath of 'path'.
+    Raises:
+        PackitException, if 'subpath' is not a subpath of 'path'.
+    """
+    if not subpath.resolve().is_relative_to(path.resolve()):
+        raise PackitException(
+            f"Sync files: Illegal path! {subpath} is not in the subpath of {path}."
+        )
+    return subpath
 
-    def __repr__(self):
-        return f"SyncFilesItem(src={self.src}, dest={self.dest})"
 
-    def __eq__(self, other: object) -> bool:
-        if not isinstance(other, SyncFilesItem):
-            raise NotImplementedError()
+class SyncFilesItem:
+    """Some files to sync to destination
 
-        return self.src == other.src and self.dest == other.dest
+    Think about this as a wrapper around 'rsync'.
 
+    Attributes:
+        src: List of paths to sync.
+        dest: Destination to sync to.
+        mkpath: Create the destination's path component.
+    """
 
-class RawSyncFilesItem(NamedTuple):
-    src: Path
-    dest: Path
-    # when dest is specified with trailing slash, it is meant to be a dir
-    dest_is_dir: bool
+    def __init__(
+        self,
+        src: Sequence[Union[str, Path]],
```

Comment:
I am maybe going to quickly through the code, but what is the benefit of having multiple source items?

Reply:
This is so that the class is a very simple and straightforward wrapper around `rsync`, which in it's local form has to following signature, taking multiple `SRC` arguments:

```
rsync [OPTION...] SRC... [DEST]
```

Another approach would have been to split configs like

```yaml
synced_files:
- src:
  - file1
  - file2
  dest: dest/
```

into `(file1, dest)` and `(file2, dest)` (as the previous implementation did it), but I felt that could quickly become complicated as new rsync-options are added, as we would need to take care how to correctly set those in the splits.

So yes, the code might be more complicated at places because of this, but it's consistently so, and overall I think this approach will help to avoid complexity in the future.

Reply:
Aha, I missed that you want to match the multi-source arguments in CLI as well. Thanks for the explanation.